sympy/solvers/solvers.py

```diff
@@ -361,7 +361,51 @@ def checksol(f, symbol, sol=None, **flags):
     # TODO: improve solution testing
 
 
-def check_assumptions(expr, against=None, failing_assumption=False, **assumptions):
+def failing_assumptions(expr, against=None, **assumptions):
+    """
+    Helper for ``check_assumptions``.
+
+    Returns empty list `[]` if all assumptions are True, or a list
+    of  assumption(s) that the expression does not satisfy or
+    `None` if it can't conclude.
+
+    Examples
+    ========
+
+        >>> from sympy import Symbol
+        >>> from sympy.solvers.solvers import failing_assumptions
+        >>> failing_assumptions(-5, integer=False, positive=True)
+        ['integer', 'positive']
+
+        >>> x = Symbol('x', positive=True)
+        >>> y = Symbol('y')
+        >>> failing_assumptions(6*x, positive=True)
+        []
+
+        >>> failing_assumptions(y, real=True)
+    """
+    if against is not None:
+        assumptions = against.assumptions0
+
+    expr = sympify(expr)
+
+    flag = True
```

Comment:
What do you think about this (no frills version)?

```
def failing_assumptions(expr, **assumptions):
    failed = {}
    for key in list(assumptions.keys()):
        test = getattr(expr, 'is_' + key, None)
        if test is not assumptions[key]:
            failed[key] = test
    return failed  # {} or {assumption: value != desired}

def check_assumptions(expr, against=None, **assumptions):
    if against:
        assert not assumptions
        assert against.is_Symbol # ?
        assumptions = against.assumptions0
    return not bool(failing_assumptions(expr, against))

>>> x = Symbol('x', real=True, positive=True)
>>> y = Symbol('y')
>>> failing_assumptions(6*x + y, real=True, positive=True)
{'real': None, 'positive': None}
>>> failing_assumptions(6*x + y, **x.assumptions0) == {
    'real': None,
    'imaginary': None,
    'complex': None,
    'hermitian': None,
    'positive': None, 
    'nonpositive': None, 
    'nonnegative': None, 
    'nonzero': None, 
    'negative': None, 
    'zero': None}
>>> failing_assumptions(x**2 - 1, positive=True)
{'positive': None}
>>> failing_assumptions(-2*x - 6, real=True, positive=True)
{'positive': False}
>>> failing_assumptions(x**2, positive=True)
{}
```

My thought is that if you are going to go through the trouble of returning what is wrong, you might as well tell how it is wrong. Imagine this being used interactively as a tutor and the task is to create an expression that is unknown to be positive. So you call `failing_assumptions(given, positive=None)`. Let's say they give a result that is positive. The session would then report that "no, your expression is positive" without having to recompute the value. Of course in the version you presented, only True and False values are returned so the key along with the passed values are sufficient for reconstruction. But it doesn't seem unreasonable to pass in an assumption that is None (as in the example given).

Reply:
To see if we both are on the same page, what I understood is you want `failing_assumptions` to return correct value of the key along with the key instead of a key only. Is it correct?

In examples above I see if an `expr` returns `None`, all assumptions are made `None`. I am not sure but I don't think this should be done because `None` comes out when atleast one (or maybe all)assumption cannot be concluded and rest are `True`. like:
`(6*x + y, real=True, positive=True)` in this real=True is correct but positive=True cannot be determined therfore the result is `None`. I think the result should be `{'real': True, 'positive': None}`

> without having to recompute the value

I am not clear on this.

Reply:
> What do you think about this (no frills version)?

I think it seems to be good, but the only question that comes to my mind is how effective this would be means will it be of any positive use(like to carry out any further computation), or will it just be a functionality that we have.

Reply:
> (6*x + y, real=True, positive=True) in this real=True is correct

Maybe you meant to make y real b/c as a vanilla y, the expression is not known to be real:

```
>>> x = Symbol('x', real=True, positive=True)
>>> y = Symbol('y')
>>> eq=6*x + y
>>> eq.is_real
>>> eq.is_positive
>>>
```

> if an expr returns None, all assumptions are made None

I'm not sure what you mean here. In this case, for all assumptions that y had the expression happened to have values of None, but this would not always be the case. For example, if we make y be an integer, it will have 11 assumptions. The expression `6*x + y` will have 6 of them that are None, not all of them:

```
>>> x = Symbol('x', real=True, positive=True)
>>> y = Symbol('y', integer=True)
>>> failing_assumptions(6*x + y, **y.assumptions0)
>>> failing_assumptions(6*x+y, **y.assumptions0)
{'integer': None, 'rational': None, 'noninteger': None, 'algebraic': None, 'irrational': None, 'transcendental': None}
>>> len(_), len(y.assumptions0)
(6, 11)
```

> will it be of any positive use

What is the use case that you are thinking of for the function -- it isn't explicitly state in the issue. (I gave an example of how one might use the information, e.g. with an interactive tutor program in which it would be useful to know the incorrect assumption and value so it can be reported to the user. Still...this seems of limited use. That's why I'm interested in seeing how you imagine such functionality to be used.)